baseclasses/SolverHist.py

```diff
@@ -0,0 +1,210 @@
+
+                Print = key.lower() in varsToPrint
+                self.addField(Name=key, Type=type(val), Print=Print)
+            else:
+                warn(
+                    "Can't add variable: '%s', from user supplied function to Solver History as there is already a tracked variable with this name, skipping"
+                    % key
+                )
+
+    def reset(self):
+        """
+        Reset the solver history log
+        """
+        self.latestIter = 0
+
+        for key in self.hist["values"]:
+            varType = self.hist["varType"][key]
+            self.hist["values"][key] = copy.deepcopy(self.defaults["values"][varType])
+
+    def write(self, data):
+        """
+        Store the values from the latest iteration in the solver history
+        """
+
+        # Figure out if we're storing the history or not
+        if self.storeValues:
+            i = self.latestIter
+        else:
+            i = 0
+
+        self.hist["values"]["Iter"][i] = self.latestIter
+        for key, val in data.items():
+            try:
+                self.hist["values"][key][i] = val
+            except KeyError:
+                Error("Option '%-30s' is not a valid history variable." % (key))
+
+        self.latestIter += 1
+
+    def printHeader(self):
+        """
+        Print the formatted header, used during nonlinear solution process
+        """
+        header = "|"
+        for var in self.varsToPrint:
+            if var in self.hist["nameFormat"]:
+                header += " " + self.hist["nameFormat"][var].format(var) + " |"
+
+        print("+" + "-" * (len(header) - 2) + "+")
+        print(header)
+        print("+" + "-" * (len(header) - 2) + "+")
+
+    def printData(self, ind=None):
+        """
+        Print a selection of the history in the correctly formatted manner, used for printing convergence history during
+        nonlinear solution process
+        """
+        if ind == None:
+            ind = [self.latestIter - 1]
+        if type(ind) is not list:
+            ind = [ind]
+        if not self.storeValues and (len(ind) > 1 or all(i not in [0, -1] for i in ind)):
+            warn(
+                'Previous iterations unavailable as py option "writeConvergenceHist" is not enabled. Printing latest iteration'
+            )
+            ind = [self.latestIter - 1]
+
+        for iter in ind:
+            if iter < self.latestIter:
+                line = "|"
+                for var in self.varsToPrint:
+                    if var in self.hist["nameFormat"]:
+                        if self.hist["varType"][var] in ["float", "int"] and np.isnan(self.hist["values"][var][iter]):
+                            line += " " + self.hist["nameFormat"][var].format("-") + " |"
+                        else:
+                            line += " " + self.hist["valueFormat"][var].format(self.hist["values"][var][iter]) + " |"
+                print(line)
+
+    def clip(self):
```

Comment:
What are the "unused" entries?

Reply:
This is a consequence of the fact that I pre-allocate the storage arrays and is why I want to change to storing in lists. For each stored variable, an array of nan's gets initially created, with a length equal to the `maxiter` option, so if ADflow (just as an example) converges before it hits it's iteration limit, you'll still have a bunch of nan's at the end of your arrays, this method just removes those.

Reply:
Ohh I see, makes sense